### PR TITLE
avoid using invalid sessions

### PR DESCRIPTION
**Pull request by yschimke**
These checks are a bit redundant, and originally came from the conscrypt branch.  We should avoid using the session if there was a failure.

I think I'm checking effectively the same thing three times.

### CODE REVIEW

**Comment by yschimke** [comment]
relates to https://github.com/square/okhttp/issues/3719

**Comment by yschimke** [comment]
@swankjesse apologies about the build failures, this particular change really depends on the SSL provider you use, so local IDE testing is tricky.  I'll fix the tests to avoid triggering these edge cases incidentally.

It will need careful review for such a small change, so we might just want to wait for the conscrypt branch to land to fix it.

File: okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java [hunk 1]
```diff
@@ -876,64 +890,46 @@ private void noRecoveryFromErrorWithRetryDisabled(ErrorCode errorCode) throws Ex
...
+    List<String> logs = http2Handler.takeAll();
+    assertThat("log count", logs.size(), anyOf(equalTo(21), equalTo(22)));
+    assertThat("header logged", firstFrame(logs, "HEADERS"), containsString("HEADERS       END_HEADERS"));
+    assertThat("data logged", firstFrame(logs, "DATA"), containsString("0 DATA          END_STREAM"));
+  }
 
-    assertEquals(1, handler.headerFrameCount);
-    assertEquals(Collections.singletonList(0), handler.dataFrames);
+  private String firstFrame(List<String> logs, String type) {
+    for (String l: logs) {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java]: nit: I really dislike `l` as a variable name because it looks like `i` and `I` and `1` and `|`. Perhaps `log` ?

File: okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java [hunk 2]
```diff
@@ -876,64 +890,46 @@ private void noRecoveryFromErrorWithRetryDisabled(ErrorCode errorCode) throws Ex
...
+      if (l.contains(type)) {
+        return l;
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java]: (this in particular looks like `return 1;`

**Comment by swankjesse** [comment]
Without having spent much time investigating this, I’m surprised that this SSL stuff is interfering with HTTP/2 framing. Those should be more independent?

File: okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java (hunk 1 above)
Comment (yschimke) [reply okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java]: My bad, I've been guilty of this before.

**Comment by yschimke** [comment]
Yeah, I wouldn't rush this in.  But it would be good to get right before Conscrypt 1.0 is released.  It seems like other providers can behave differently for sessions during connection and when errors/disconnects happen.  So I think we should tighten up OkHttp here.

**Comment by yschimke** [comment]
Can you explain more about the HTTP/2 framing, it's possible my fixes in the tests are masking a real problem with these SSL providers.

**Comment by swankjesse** [comment]
If we're doing things right TLS and HTTP/2 should be very independent of one another. Changing TLS setup shouldn't be able to impact framing because framing shouldn't even know TLS is present.

Is this two unrelated changes in one PR? Or is there a link between the TLS change and the test change that I'm missing?

**Comment by yschimke** [comment]
I think the change is broken as is, let me simplify it further and try to avoid impacting any tests.

**Comment by yschimke** [comment]
@swankjesse is this acceptable in this minimal form?

**Comment by swankjesse** [comment]
Yes, perfect. I’d incorrectly inferred that the tests and the fix were related.
